### ISSUE TITLE
Find a better way to convert `AlgonautError`s that result from a 404 HTTP status code

### ISSUE THREAD

**Issue by AlterionX** [manuelmauro/algonaut#126]
**Is your feature request related to a problem? Please describe.**
It's kind of difficult to get the status code of the error for `AlgonautError::RequestError`. Namely:

`AlgonautError::Request(RequestError { details: RequestErrorDetails::Http { status: 404, ... } , ...})`

It's nested fairly deep within the error struct, and I feel like this should be a relatively common usage of the error if people attempt to recover from errors.

**Describe the solution you'd like/Describe alternatives you've considered**
I'm uncertain about what I actually want to see. So I'd like to talk about it.
I'd probably want a combination of the following ideas:
- Understand 404 errors from the indexer when attempting to fetch information about a transaction and return None.
  - Inflexible, potentially not what's intended for a direct translation of the API. Some people might want 404 to be an error.
- Create a method, similar to diesel's `optional` method on queries, that will do this instead of changing the function signature.
  - Simple, but highly specialized. Possibly inaccurate.
- Add a method to `RequestError` to check for 404s.
  - Specific, but I'm not sure any other codes are really worth checking. Will need to revisit the REST API specifications.
- Simply create a method on `AlgonautError`/`RequestError`/`RequestErrorDetails` to return `Option<u16>` when asking for the HTTP status code.
  - It's mostly a question of granularity between this and the previous entry as well as how much parsing is required.
I'm sure there are better solutions that what I've thought of, but this is what I have.

So far, I've been dealing with it by creating a function to call instead of the default API's function per project.

**Additional context**
Possibly relates to #99 

EDIT: I'm looking specifically at `Indexer::transaction_info`.

**Comment by AlterionX** [comment]
Might also be that 404 doesn't mean what I think it means. I've been reading the documentation, but it doesn't really specify.

**Comment by ivanschuetz** [comment]
I see several (partly independent) issues:

1) Nesting: I agree that it should be improved. Ideally there shouldn't be any nesting to access the primary / semantic cause of the error.

2) Whether we generally want to return HTTP status codes to the user or map to semantic errors (meaning depending on the specific request). Probably only 2-3 relevant codes (400, 404 and maybe 503).

3) If mapping, whether we want to return different error enums per response / response groups where the meanings are the same.

4) Whether we want to map "not found" to `None` - so not an error (I tend to agree with this). I'd probably do it in the high level wrappers.

5) https://github.com/manuelmauro/algonaut/issues/99 Is mostly unrelated. It's about the confusion caused by public-facing error types like `TransactionError` and `AlgonautError` (might be the only ones), which I think should be solved by renaming `AlgonautError` into something more specific (so it doesn't look like it's the only / main public error type)

**Comment by AlterionX** [comment]
I'm good with 1 & 2.

3. Sounds even if we do, I'd like the enum to have something like a `is_404` method on it to quickly verify. Similar to `is_some` or `is_none` for `Option`.
4. Most of my input is assumed to be invalid until validated with the indexer so I'm also assuming that the input is invalid. I'd just like the option to pick -- we don't need to change default behavior. (I'll just reference `optional` from `diesel` again.)
5. Noted.

**Comment by ivanschuetz** [comment]
Ok, I think that we can add the HTTP statuses (404 and 400) directly to [AlgonautError](https://github.com/manuelmauro/algonaut/blob/main/src/error.rs#L5). I'd ideally prefer to use semantic errors (e.g. "Not found" or ["Developer API not enabled"](https://developer.algorand.org/docs/rest-apis/algod/v2/#post-v2tealcompile) instead of 404. But given that the documentation (as you also indicated) is a bit incomplete, and might be unstable, this seems reasonable for now. We can continue evolving this.

As to the conversion to `Option`, it's true that in the time that it takes to index, the "not found" probably shouldn't be considered an error, but we don't have information of whether this is the cause and I'd prefer to consistently treat it always in the same way. This keeps the SDK lean and easier to reason about. A generic query builder like diesel is IMO a bit different in that there's no "normal" state where you expect the results to exist, but it entirely depends on what you're using it for. The trait to convert to `Option` should also be easy to add outside.

**Comment by AlterionX** [comment]
Sounds good to me. I'll get started on a PR for it, if that's alright.

To confirm: I'd be tweaking the `AlgonautError` enum to have two additional variants (e.g. `Request404`/`Request400`) with the same structure as the current `Request`? I think I should I rename `Request` as well, just thinking of a name right now. I'll also be adding a few functions that can quickly verify if an error is of a certain variant for all variants, such as the `is_404` mentioned above along with things like `is_bad_token`.

**Comment by ivanschuetz** [comment]
Hmm, what about we add convenience that returns the complete error - e.g. `_404() -> Option<HttpError>`? This would address the nesting (to be used with `is_some()` for the "is")?

The current implementation:

```rust
#[derive(Error, Debug, PartialEq, Eq)]
#[error("{:?}, {}", url, details)]
pub struct RequestError {
    pub url: Option<String>,
    pub details: RequestErrorDetails,
}

#[derive(Error, Debug, PartialEq, Eq)]
pub enum RequestErrorDetails {
    /// Http call error with optional message (returned by remote API)
    #[error("Http error: {}, {}", status, message)]
    Http { status: u16, message: String },
    /// Timeout
    #[error("Timeout connecting to the server.")]
    Timeout,
    /// Client generated errors (while e.g. building request or decoding response)
    #[error("Client error: {}", description)]
    Client { description: String },
}
```

is not incorrect (only inconvenient). Flattening it would lead to repetition.

**Comment by ivanschuetz** [comment]
Tbh, I'm personally more on the "rustic" side of things and don't see that much of a problem with this:

```rust
match error {
    Request(RequestError { details: Http { status: 404, message: _ }, .. }) => todo!(),
    _ => todo!()
}
```

It's a bit "low level", but it seems that we care only about recoverable errors like the indexer queries, and adding convenience to the whole API because of that might be overkill. Do we need this for 400?  The client generated errors? How to handle timeouts? Too many things to think about, for a very small benefit IMO. Also, this code doesn't have to be written frequently. It's normally in only one place in the app + you can easily add your own convenience / extensions to `AlgonautError` if really needed. Interested in reading more opinions though.

**Comment by ivanschuetz** [comment]
Anyway, to conclude my part, I'm ok either leaving it like it is or adding some convenience. Given that the use cases beyond 404 seem unclear, I'd add a check just for that. We can add more later if needed.

My enum idea can be ignored. I think that it requires a bit more advanced modeling of the errors, which would include specifying all the possible causes and using different types per requests / request groups. With just the status codes it would be a rather bad design.

**Comment by AlterionX** [comment]
Sorry, been a bit busy, so late response.

It's not unmanageable, but it's kind of hard to read. There's just a bunch of noise, lol. Like, compare
```rs
Request(RequestError { details: Http { status: 404, message: _ }, .. })
```
with
```rs
error.is_404()
```
Of course, if you match for multiple errors, the full match might make more sense, but I've found that 404 is essentially the only error I end up responding to.

Also, I think part of what makes Rust + its ecosystem so nice is that annoying things gets a feature update 8 times out of 10. We could just use match statements. But we have `is_some`. We could just unwrap `Result`s manually. But we have `?`.

That, and that the said feature attempts to be "correct", whatever that means.

Well, it'll also be nice to have that enum thing after the API stabilizes a bit more, and we get some more info on how errors in the API should be handled.

TL;DR

I'll just stick a small `is_404` function onto the error, then.
